cf-check/validate.c

```diff
@@ -0,0 +1,515 @@
+
+static void ValidateStateLastseen(ValidatorState *state)
+{
+    const LastSeenState ls = state->lastseen;
+    StringMap *const hostkey_to_address = ls.hostkey_to_address;
+    StringMap *const address_to_hostkey = ls.address_to_hostkey;
+    StringSet *const quality_outgoing_hostkeys = ls.quality_outgoing_hostkeys;
+    StringSet *const quality_incoming_hostkeys = ls.quality_incoming_hostkeys;
+
+    assert(state != NULL);
+
+    {
+        MapIterator iter = MapIteratorInit(hostkey_to_address->impl);
+        MapKeyValue *current_item;
+        while ((current_item = MapIteratorNext(&iter)) != NULL)
+        {
+            const char *hostkey = current_item->key;
+            if (!ValidateHostkey(state, hostkey))
+            {
+                continue;
+            }
+            const char *address = current_item->value;
+            if (!ValidateAddress(state, address))
+            {
+                continue;
+            }
+            const char *lookup = StringMapGet(address_to_hostkey, address);
+            if (lookup == NULL)
+            {
+                ValidationError(
+                    state, "Missing address entry for '%s'", address);
+            }
+            else if (!StringSafeEqual(hostkey, lookup))
+            {
+                ValidationError(
+                    state,
+                    "Bad hostkey->address->hostkey reverse lookup '%s' != '%s'",
+                    hostkey,
+                    lookup);
+            }
+        }
+    }
+    {
+        MapIterator iter = MapIteratorInit(address_to_hostkey->impl);
+        MapKeyValue *current_item;
+        while ((current_item = MapIteratorNext(&iter)) != NULL)
+        {
+            const char *address = current_item->key;
+            if (!ValidateAddress(state, address))
+            {
+                continue;
+            }
+            const char *hostkey = current_item->value;
+            if (!ValidateHostkey(state, hostkey))
+            {
+                continue;
+            }
+            const char *lookup = StringMapGet(hostkey_to_address, hostkey);
+            if (lookup == NULL)
+            {
+                ValidationError(
+                    state, "Missing hostkey entry for '%s'", hostkey);
+            }
+            else if (!StringSafeEqual(address, lookup))
+            {
+                ValidationError(
+                    state,
+                    "Bad address->hostkey->address reverse lookup '%s' != '%s'",
+                    address,
+                    lookup);
+            }
+        }
+    }
+    {
+        StringSetIterator iter =
+            StringSetIteratorInit(quality_incoming_hostkeys);
+        const char *hostkey;
+        while ((hostkey = StringSetIteratorNext(&iter)) != NULL)
+        {
+            if (StringMapGet(hostkey_to_address, hostkey) == NULL)
+            {
+                ValidationError(
+                    state,
+                    "Missing hostkey from quality in entry '%s'",
```

Comment:
This message sounds quite confusing to me.

Reply:
Clarified a little bit by adding a hyphen: `quality-in`.